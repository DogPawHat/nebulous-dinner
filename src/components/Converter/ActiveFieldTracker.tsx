import { Component, ReactNode, Key } from 'react';


export type IMakeActive = (key: Key) => () => void;
export type IIsActive = (key: Key) => boolean;
export type IActiveFieldTrackerRenderProp = (
  isActive: IIsActive,
  makeActive: IMakeActive
) => ReactNode;


interface IActiveFieldTrackerProps {
  defaultKey: Key;
  render: IActiveFieldTrackerRenderProp
}

interface IActiveFieldTrackerState {
  activeKey: Key;
}


class ActiveFieldTracker extends Component<IActiveFieldTrackerProps, IActiveFieldTrackerState> {

  constructor(props: IActiveFieldTrackerProps) {
    super(props);
    this.state = {
      activeKey: props.defaultKey
    }
  }

  public makeActive: IMakeActive = (key: Key) => () => {
    this.setState({
      activeKey: key
    })
  }

  public isActive: IIsActive = (key: Key) => key === this.state.activeKey;

  public render() {
    return this.props.render(this.isActive, this.makeActive);
  }
}

export default ActiveFieldTracker;
