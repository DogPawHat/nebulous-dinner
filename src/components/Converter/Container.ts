import { connect, MapStateToProps } from 'react-redux';

import { IState, actions } from '../../ducks';
import { IConverterStateProps, IConverterOwnProps, IConverterDispatchProps } from './interfaces';

const mapStateToProps: MapStateToProps<
  IConverterStateProps,
  IConverterOwnProps,
  IState
> = state => ({
  defaultValue: state.sending,
  defaultCurrency: 'EUR',
  eurToGbpRate: state.rate,
  fee: state.fee
});

const mapDispactchToProps: IConverterDispatchProps = {
  notifyUpdate: actions.updateCurrency
};

const enhance = connect(mapStateToProps, mapDispactchToProps);

export default enhance;
