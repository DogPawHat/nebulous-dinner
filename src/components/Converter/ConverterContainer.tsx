import styled, { css } from 'react-emotion';

export interface IConverterActiveChildProps {
  active: boolean;
}

const activeStyle = css`
  transform: translateZ(5px);
  :not(:first-child) {
    transform: translateY(-1px) transformZ(5px);
  }
`;

const baseStyle = css`
  :first-child {
    border-top-right-radius: 4px;
    border-top-left-radius: 4px;
  }

  :last-child {
    border-bottom-right-radius: 4px;
    border-bottom-left-radius: 4px;
  }

  :not(:first-child) {
    transform: translateY(-1px);
  }
`;

export const ConverterContainerChild =
  ({ active }: IConverterActiveChildProps) => css`
    ${baseStyle}
    ${active ? activeStyle : undefined}
  `;

const ConverterContainer = styled('div')`
  transform-style: preserve-3d;
`;

export default ConverterContainer;
