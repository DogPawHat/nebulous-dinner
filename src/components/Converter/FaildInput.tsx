import React, {
  Component,
  ChangeEventHandler
} from 'react';
import styled from 'react-emotion';
import { fontSizes, fontWeights, colors } from '../../styleUtils';

interface IInputProps {
  value?: string;
  onChange: (value: string) => void;
}

// CSS
const intergerInputClassName = 'Input__interger';
const decimalInputClassName = 'Input__decimal';
const dotClassName = 'Input__dot';

const InputRoot = styled('div')`
  display: flex;
  flex-flow: row nowrap;
  align-items: baseline;
  height: 2.375rem;
  .${intergerInputClassName}, .${decimalInputClassName}, .${dotClassName} {
    font-weight: ${fontWeights.medium};
    font-size: ${fontSizes.large};
    flex: 0 1 auto;
  }
  .${intergerInputClassName}, .${decimalInputClassName} {
    border: none;
    background: transparent;
  }
  .${intergerInputClassName} {
    font-size: ${fontSizes.huge};
    color: ${colors.black};
  }
  .${decimalInputClassName}, .${dotClassName} {
    font-size: ${fontSizes.large};
    color: ${colors.grey};
  }
`;

const largeFigureRegx = /^[£€].?[1-9][\d,]*/;
const smallFigureRegx = /\d\d$/;
const wholeFigureRegX = /^([£€].?[1-9][\d,]*)\.(\d\d)$/;

const checkValue = (value: string) => wholeFigureRegX.test(value);

class Input extends Component<IInputProps> {
  get integerValue() {
    const { value } = this.props;

    if (!value || !checkValue(value)) {
      return '0';
    }

    const match = value.match(largeFigureRegx);

    if (!match || match.length !== 1) {
      return '0';
    }

    return match[0];
  }

  get decimalValue() {
    const { value } = this.props;

    if (!value || !checkValue(value)) {
      return '0';
    }

    const match = value.match(smallFigureRegx);

    if (!match || match.length !== 1) {
      return '0';
    }

    return match[0];
  }

  public onIntegerValueChange: ChangeEventHandler<HTMLInputElement> = event => {
    const { value } = event.target;

    const newValue = `${value}.${this.decimalValue}`;

    this.props.onChange(newValue);
  };

  public onDecimalValueChange: ChangeEventHandler<HTMLInputElement> = event => {
    const { value } = event.target;

    const newValue = `${this.integerValue}.${value}`;

    this.props.onChange(newValue);
  };

  public render() {
    return (
      <InputRoot>
        <span
          contentEditable={true}
          onInput={this.onIntegerValueChange}
          className={intergerInputClassName}
        >{this.integerValue}</span>
        <span className={dotClassName}>.</span>
        <span
          contentEditable={true}
          onInput={this.onDecimalValueChange}
          className={decimalInputClassName}
        >{this.decimalValue}</span>
      </InputRoot>
    );
  }
}

export default Input;
