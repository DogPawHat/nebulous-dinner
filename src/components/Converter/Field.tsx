import React, { SFC, HTMLAttributes } from 'react';
import styled, { css } from 'react-emotion';
import { fontSizes, fontWeights, colors } from 'src/styleUtils';
import color from 'color';
import InputMasker, { IInputMaskerRenderProp } from './InputMasker';
import Input from './Input';

// Interfaces
export interface IConverterFieldProps extends HTMLAttributes<HTMLDivElement> {
  description: string;
  active: boolean;
  position?: 'top' | 'bottom';
}

interface IConverterFieldRootProps {
  active: boolean;
}

type IConverterField = SFC<IConverterFieldProps>;

// CSS classes
const activeBg = ({ active }: IConverterFieldRootProps) => css`
  background-color: ${active
    ? color(colors.blue)
        .alpha(0.3)
        .string()
    : colors.white};
`;

const activeColor = ({ active }: IConverterFieldRootProps) => css`
  color: ${active ? colors.blue : colors.grey};
`;

const activeTransform = ({ active }: IConverterFieldRootProps) => css`
  color: ${active ? colors.blue : colors.grey};
`;

const activeBorderColor = ({ active }: IConverterFieldRootProps) => css`
  transform: ${active ? 'translateZ(5px)' : 'translateZ(0)'};
`;

const descriptionClassName = 'Converter__description';

// Styled Components
const FieldRoot = styled('div')`
  ${activeBg} ${activeBorderColor} ${activeTransform}
  padding: 0 24px;
  border-width: 1px;
  border-style: solid;
  .${descriptionClassName} { 
    line-height: 2.375rem;
    font-weight: ${fontWeights.medium};
    ${activeColor}
    font-size: ${fontSizes.body};
    text-transform: uppercase;
    line-height: 38px;
  }
`;


const maskerRenderProp: IInputMaskerRenderProp = (value, onChange) => (
  <Input value={value} onChange={onChange} />
)

// Main
const Field: IConverterField = ({ description, ...otherProps }) => (
  <FieldRoot {...otherProps}>
    <p className={descriptionClassName}>{description}</p>
    <InputMasker defaultValue={2000} render={maskerRenderProp} />
  </FieldRoot>
);

export default Field;
