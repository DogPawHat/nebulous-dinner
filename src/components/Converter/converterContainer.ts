// The converter file has some tricky CSS
// Esssentally to get the layout right, this is what I had to do:


import { css } from 'react-emotion';

export interface IConverterActiveChildProps {
  active: boolean;
}

export const converterContainerClassName = 'ConverterContainer';

// Create a style for the active (blue higlighted) field, ensuring it
// has a border on 4 sides (actual colours are applied in the Field.tsx file)
// Also ensure the next field has no top border (prevents double border)
const activeStyle = css`
  .${converterContainerClassName} & {
    z-index: 5;
  }

  .${converterContainerClassName} &:first-child + * {
    border-top-style: none;
  }
`;

// Create styleds for the other, no active field (or fields)
// The way this works is:
// 1. have the last child have a border on all four sides
// 2. Have the rest only have a border on the top, left and right
// This gives the collection of fields a table like border-collape appeance
const nonActiveStyle = css`
  .${converterContainerClassName} &:not(:last-child) {
    border-bottom-style: none;
  }
`;

const baseStyle = css`
  .${converterContainerClassName} &:first-child {
    border-top-right-radius: 4px;
    border-top-left-radius: 4px;
  }

  .${converterContainerClassName} &:last-child {
    border-bottom-right-radius: 4px;
    border-bottom-left-radius: 4px;
  }
`;

export const converterContainerChildClassName =
  ({ active }: IConverterActiveChildProps) => css`
    ${baseStyle}
    ${active ? activeStyle : nonActiveStyle}
  `;
