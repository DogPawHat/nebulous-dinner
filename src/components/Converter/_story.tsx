import { withInfo } from '@storybook/addon-info';
import { storiesOf } from '@storybook/react';
import React from 'react';

import Converter from '.';
import { css } from 'react-emotion';

const sizeAndMargin = css`
  width: 35.25rem;
  margin: 3.3rem;
`;

storiesOf('Converter', module).add(
  'default',
  withInfo({ inline: true })(() => <Converter
    defaultCurrency='EUR'
    defaultValue={2000}
    eurToGbpRate={0.86022}
    className={sizeAndMargin} />)
);
