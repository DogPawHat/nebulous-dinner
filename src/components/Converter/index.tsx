import React, {SFC} from 'react';
import {
  converterContainerClassName,
  converterContainerChildClassName,
} from './converterContainer';
import Field from './Field';
import ActiveFieldTracker, {
  IActiveFieldTrackerRenderProp
} from './ActiveFieldTracker';
import styled, { cx } from 'react-emotion';

interface IConverterProps {
  className?: string
}

type IConverter = SFC<IConverterProps>;

const FieldChild = styled(Field)`
  ${converterContainerChildClassName}
`;

// Render prop for converter
const renderFields: IActiveFieldTrackerRenderProp = (isActive, makeActive) => (
  <>
    <FieldChild
      key="field_1"
      active={isActive('field_1')}
      onClick={makeActive('field_1')}
      description="you send"
    />
    <FieldChild
      key="field_2"
      active={isActive('field_2')}
      onClick={makeActive('field_2')}
      description="reciver gets"
    />
  </>
);

const Converter: IConverter = ({className}) => (
  <div className={cx(className, converterContainerClassName)}>
    <ActiveFieldTracker defaultKey="field_1" render={renderFields} />
  </div>
);

export default Converter;
