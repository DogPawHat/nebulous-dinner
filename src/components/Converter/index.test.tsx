
import React from 'react'
import {render, cleanup} from 'react-testing-library'
// this add custom jest matchers from jest-dom
import 'jest-dom/extend-expect'

import Converter from '.';

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup)

test('Converter test', () => {
  // Arrange
  const {queryByText} = render(<Converter
    defaultCurrency='EUR'
    defaultValue={2000}
    eurToGbpRate={0.86022}
  />);

  // Assert
  expect(queryByText('€ 2000.00')).not.toBeNull();
  expect(queryByText('£ 1,717.94')).not.toBeNull();
});
