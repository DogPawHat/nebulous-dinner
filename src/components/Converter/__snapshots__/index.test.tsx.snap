// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Converter euro to pound 1`] = `
<div>
  <div
    class="ConverterContainer"
  >
    <div
      class="css-wwdlxm"
    >
      <p
        class="Converter__description"
      >
        you send
      </p>
      <input
        class="Converter__input"
        data-testid="input_EUR"
        value="2,000.00"
      />
    </div>
    <div
      class="css-kl0by6"
    >
      <p
        class="Converter__description"
      >
        reciver gets
      </p>
      <input
        class="Converter__input"
        data-testid="input_GBP"
        value="1,717.94"
      />
    </div>
  </div>
</div>
`;

exports[`Converter euro to pound dynamic 1`] = `
<div>
  <div
    class="ConverterContainer"
  >
    <div
      class="css-wwdlxm"
    >
      <p
        class="Converter__description"
      >
        you send
      </p>
      <input
        class="Converter__input"
        data-testid="input_EUR"
        value="3,000.00"
      />
    </div>
    <div
      class="css-kl0by6"
    >
      <p
        class="Converter__description"
      >
        reciver gets
      </p>
      <input
        class="Converter__input"
        data-testid="input_GBP"
        value="2,578.16"
      />
    </div>
  </div>
</div>
`;

exports[`Converter euro to pound dynamic with different fee 1`] = `
<div>
  <div
    class="ConverterContainer"
  >
    <div
      class="css-wwdlxm"
    >
      <p
        class="Converter__description"
      >
        you send
      </p>
      <input
        class="Converter__input"
        data-testid="input_EUR"
        value="3,000.00"
      />
    </div>
    <div
      class="css-kl0by6"
    >
      <p
        class="Converter__description"
      >
        reciver gets
      </p>
      <input
        class="Converter__input"
        data-testid="input_GBP"
        value="2,570.66"
      />
    </div>
  </div>
</div>
`;

exports[`Converter euro to pound dynamic with different rate 1`] = `
<div>
  <div
    class="ConverterContainer"
  >
    <div
      class="css-wwdlxm"
    >
      <p
        class="Converter__description"
      >
        you send
      </p>
      <input
        class="Converter__input"
        data-testid="input_EUR"
        value="3,000.00"
      />
    </div>
    <div
      class="css-kl0by6"
    >
      <p
        class="Converter__description"
      >
        reciver gets
      </p>
      <input
        class="Converter__input"
        data-testid="input_GBP"
        value="2,669.51"
      />
    </div>
  </div>
</div>
`;

exports[`Converter euro to pound with different fee 1`] = `
<div>
  <div
    class="ConverterContainer"
  >
    <div
      class="css-wwdlxm"
    >
      <p
        class="Converter__description"
      >
        you send
      </p>
      <input
        class="Converter__input"
        data-testid="input_EUR"
        value="2,000.00"
      />
    </div>
    <div
      class="css-kl0by6"
    >
      <p
        class="Converter__description"
      >
        reciver gets
      </p>
      <input
        class="Converter__input"
        data-testid="input_GBP"
        value="1,710.44"
      />
    </div>
  </div>
</div>
`;

exports[`Converter euro to pound with different rate 1`] = `
<div>
  <div
    class="ConverterContainer"
  >
    <div
      class="css-wwdlxm"
    >
      <p
        class="Converter__description"
      >
        you send
      </p>
      <input
        class="Converter__input"
        data-testid="input_EUR"
        value="2,000.00"
      />
    </div>
    <div
      class="css-kl0by6"
    >
      <p
        class="Converter__description"
      >
        reciver gets
      </p>
      <input
        class="Converter__input"
        data-testid="input_GBP"
        value="1,778.84"
      />
    </div>
  </div>
</div>
`;

exports[`Converter euro to pound with different value 1`] = `
<div>
  <div
    class="ConverterContainer"
  >
    <div
      class="css-wwdlxm"
    >
      <p
        class="Converter__description"
      >
        you send
      </p>
      <input
        class="Converter__input"
        data-testid="input_EUR"
        value="3,000.00"
      />
    </div>
    <div
      class="css-kl0by6"
    >
      <p
        class="Converter__description"
      >
        reciver gets
      </p>
      <input
        class="Converter__input"
        data-testid="input_GBP"
        value="2,578.16"
      />
    </div>
  </div>
</div>
`;

exports[`Converter pound to euro 1`] = `
<div>
  <div
    class="ConverterContainer"
  >
    <div
      class="css-kl0by6"
    >
      <p
        class="Converter__description"
      >
        you send
      </p>
      <input
        class="Converter__input"
        data-testid="input_EUR"
        value="2,322.49"
      />
    </div>
    <div
      class="css-wwdlxm"
    >
      <p
        class="Converter__description"
      >
        reciver gets
      </p>
      <input
        class="Converter__input"
        data-testid="input_GBP"
        value="2,000.00"
      />
    </div>
  </div>
</div>
`;

exports[`Converter pound to euro dynamic 1`] = `
<div>
  <div
    class="ConverterContainer"
  >
    <div
      class="css-kl0by6"
    >
      <p
        class="Converter__description"
      >
        you send
      </p>
      <input
        class="Converter__input"
        data-testid="input_EUR"
        value="3,484.98"
      />
    </div>
    <div
      class="css-wwdlxm"
    >
      <p
        class="Converter__description"
      >
        reciver gets
      </p>
      <input
        class="Converter__input"
        data-testid="input_GBP"
        value="3,000.00"
      />
    </div>
  </div>
</div>
`;

exports[`Converter pound to euro dynamic with different fee 1`] = `
<div>
  <div
    class="ConverterContainer"
  >
    <div
      class="css-kl0by6"
    >
      <p
        class="Converter__description"
      >
        you send
      </p>
      <input
        class="Converter__input"
        data-testid="input_EUR"
        value="3,477.48"
      />
    </div>
    <div
      class="css-wwdlxm"
    >
      <p
        class="Converter__description"
      >
        reciver gets
      </p>
      <input
        class="Converter__input"
        data-testid="input_GBP"
        value="3,000.00"
      />
    </div>
  </div>
</div>
`;

exports[`Converter pound to euro dynamic with different rate 1`] = `
<div>
  <div
    class="ConverterContainer"
  >
    <div
      class="css-kl0by6"
    >
      <p
        class="Converter__description"
      >
        you send
      </p>
      <input
        class="Converter__input"
        data-testid="input_EUR"
        value="3,365.75"
      />
    </div>
    <div
      class="css-wwdlxm"
    >
      <p
        class="Converter__description"
      >
        reciver gets
      </p>
      <input
        class="Converter__input"
        data-testid="input_GBP"
        value="3,000.00"
      />
    </div>
  </div>
</div>
`;

exports[`Converter pound to euro with different fee 1`] = `
<div>
  <div
    class="ConverterContainer"
  >
    <div
      class="css-kl0by6"
    >
      <p
        class="Converter__description"
      >
        you send
      </p>
      <input
        class="Converter__input"
        data-testid="input_EUR"
        value="2,314.99"
      />
    </div>
    <div
      class="css-wwdlxm"
    >
      <p
        class="Converter__description"
      >
        reciver gets
      </p>
      <input
        class="Converter__input"
        data-testid="input_GBP"
        value="2,000.00"
      />
    </div>
  </div>
</div>
`;

exports[`Converter pound to euro with different rate 1`] = `
<div>
  <div
    class="ConverterContainer"
  >
    <div
      class="css-kl0by6"
    >
      <p
        class="Converter__description"
      >
        you send
      </p>
      <input
        class="Converter__input"
        data-testid="input_EUR"
        value="2,243.00"
      />
    </div>
    <div
      class="css-wwdlxm"
    >
      <p
        class="Converter__description"
      >
        reciver gets
      </p>
      <input
        class="Converter__input"
        data-testid="input_GBP"
        value="2,000.00"
      />
    </div>
  </div>
</div>
`;

exports[`Converter pound to euro with different value 1`] = `
<div>
  <div
    class="ConverterContainer"
  >
    <div
      class="css-kl0by6"
    >
      <p
        class="Converter__description"
      >
        you send
      </p>
      <input
        class="Converter__input"
        data-testid="input_EUR"
        value="3,484.98"
      />
    </div>
    <div
      class="css-wwdlxm"
    >
      <p
        class="Converter__description"
      >
        reciver gets
      </p>
      <input
        class="Converter__input"
        data-testid="input_GBP"
        value="3,000.00"
      />
    </div>
  </div>
</div>
`;
